src/outagesim/backup_reliability.jl

```diff
@@ -0,0 +1,925 @@
+            batt_discharge_efficiency = p.s.storage.attr[b].discharge_efficiency
+        end
+            
+        batt_kwh = get(d["Storage"], "size_kwh", 0)
+        batt_kw = get(d["Storage"], "size_kw", 0)
+        init_soc = get(d["Storage"], "year_one_soc_series_pct", [])
+        starting_batt_soc_kwh = init_soc .* batt_kwh
+    end
+    if microgrid_only && !Bool(get(d, "storage_upgraded", false))
+        batt_kwh = 0
+        batt_kw = 0
+        init_soc = []
+    end
+
+    diesel_kw = 0
+    if "Generator" in keys(d)
+        diesel_kw = get(d["Generator"], "size_kw", 0)
+    end
+    if microgrid_only
+        diesel_kw = get(d, "Generator_mg_kw", 0)
+    end
+
+    #Add backup reliability values
+    max_outage_duration = p.s.backup_reliability.max_outage_duration
+
+    if max_outage_duration == 0
+        return [] #TODO add zero results
+    end
+    
+
+    #If gen capacity is 0 then base on diesel_kw
+    #If num_gens is zero then either set to 1 or base on ceiling(diesel_kw / gen_capacity)
+    gen_oa = p.s.backup_reliability.gen_oa
+    gen_fts = p.s.backup_reliability.gen_fts
+    gen_ftr = p.s.backup_reliability.gen_ftr
+    num_gens = p.s.backup_reliability.num_gens
+    gen_capacity = p.s.backup_reliability.gen_capacity
+    num_battery_bins = p.s.backup_reliability.num_battery_bins
+    microgrid_only = p.s.backup_reliability.microgrid_only
+    if gen_capacity < 0.1
+        if num_gens <= 1
+            gen_capacity = diesel_kw
+            num_gens = 1
+        else
+            gen_capacity = diesel_kw / num_gens
+        end
+    elseif num_gens == 0
+        num_gens = ceil(Int, diesel_kw / gen_capacity)
+    end
+    
+    #No reliability calculations if no generators
+    if gen_capacity < 0.1
+        return []
+    elseif batt_kw < 0.1
+        critical_loads_kw = p.s.electric_load.critical_loads_kw
+        return [survival_over_time_gen_only(critical_loads_kw, gen_oa, gen_fts, gen_ftr, num_gens, gen_capacity, max_outage_duration, marginal_survival = true),
+                survival_over_time_gen_only(critical_loads_kw, gen_oa, gen_fts, gen_ftr, num_gens, gen_capacity, max_outage_duration, marginal_survival = false)]
+
+    else
+        net_critical_loads_kw = p.s.electric_load.critical_loads_kw .- pv_kw_ac_hourly
+        return [survival_with_battery(net_critical_loads_kw, starting_batt_soc_kwh, gen_oa, gen_fts, gen_ftr, num_gens, gen_capacity, batt_kwh, batt_kw,
+                                        num_battery_bins, max_outage_duration, batt_charge_efficiency, batt_discharge_efficiency, marginal_survival = true),
+                survival_with_battery(net_critical_loads_kw, starting_batt_soc_kwh, gen_oa, gen_fts, gen_ftr, num_gens, gen_capacity, batt_kwh, batt_kw,
+                                        num_battery_bins, max_outage_duration, batt_charge_efficiency, batt_discharge_efficiency, marginal_survival = false)] 
+
+    end
+
```

Comment:
please remove extra new lines (two new lines between functions/doc strings is great and one new line is nice for legibility in the code - no need to change the spacing you have in general, but there are 4 new lines here and a couple places that you have 2 new lines between code where 1 is sufficient.)